# 1.Load data

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('ner_dataset.csv', encoding= 'unicode_escape')
data.head()

Sentence #           Word  POS Tag
0  Sentence: 1      Thousands  NNS   O
1          NaN             of   IN   O
2          NaN  demonstrators  NNS   O
3          NaN           have  VBP   O
4          NaN        marched  VBN   O

In [3]:
data.shape

(1048575, 4)

We can see that the sentences are broken into tokens in the column 'Word'. The column 'sentence #' displays the sentence number once and then prints NaN till the next sentence begins.

## 2. Extract mappings required for the neural network
To train a neural network, we will use two mappings as given below. The neural network will only take integers as input. So lets convert all the unique tokens in the corpus to its respective index.
- {token} to {token id}: address the row in embeddings matrix for the current token.
- {tag} to {tag id}: one-hot ground truth probability distribution vectors for computing the loss at the output of the network.

In [4]:
from itertools import chain
def get_dict_map(data, token_or_tag):
    tok2idx = {}
    idx2tok = {}
    
    if token_or_tag == 'token':
        vocab = list(set(data['Word'].to_list()))
    else:
        vocab = list(set(data['Tag'].to_list()))
    
    idx2tok = {idx:tok for  idx, tok in enumerate(vocab)}
    tok2idx = {tok:idx for  idx, tok in enumerate(vocab)}
    return tok2idx, idx2tok


token2idx, idx2token = get_dict_map(data, 'token')
tag2idx, idx2tag = get_dict_map(data, 'tag')

In [5]:
data['Word_idx'] = data['Word'].map(token2idx)
data['Tag_idx'] = data['Tag'].map(tag2idx) 

In [6]:
data.head()

Sentence #           Word  POS Tag  Word_idx  Tag_idx
0  Sentence: 1      Thousands  NNS   O     26203       16
1          NaN             of   IN   O     28887       16
2          NaN  demonstrators  NNS   O     29245       16
3          NaN           have  VBP   O     19764       16
4          NaN        marched  VBN   O      5249       16

## 3. Transform columns to extract sequential data
Next, lets fill NaN in 'sentence #' column using method ffill in fillna. Thereafter groupby on the sentence column to get a list of tokens and tags for each sentence.

In [7]:
data.isna().any()

Sentence #     True
Word          False
POS           False
Tag           False
Word_idx      False
Tag_idx       False
dtype: bool

In [8]:
data_fillna = data.fillna(method='ffill', axis=0)
data_group = data_fillna.groupby(['Sentence #'],as_index=False
                                )['Word', 'POS', 'Tag', 'Word_idx', 'Tag_idx'].agg(lambda x: list(x))

data_group.head()

C:\Users\Prajwal Bhamere\AppData\Local\Temp\ipykernel_10936\1778839611.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  data_group = data_fillna.groupby(['Sentence #'],as_index=False


Sentence #                                               Word  \
0      Sentence: 1  [Thousands, of, demonstrators, have, marched, ...   
1     Sentence: 10  [Iranian, officials, say, they, expect, to, ge...   
2    Sentence: 100  [Helicopter, gunships, Saturday, pounded, mili...   
3   Sentence: 1000  [They, left, after, a, tense, hour-long, stand...   
4  Sentence: 10000  [U.N., relief, coordinator, Jan, Egeland, said...   

                                                 POS  \
0  [NNS, IN, NNS, VBP, VBN, IN, NNP, TO, VB, DT, ...   
1  [JJ, NNS, VBP, PRP, VBP, TO, VB, NN, TO, JJ, J...   
2  [NN, NNS, NNP, VBD, JJ, NNS, IN, DT, NNP, JJ, ...   
3     [PRP, VBD, IN, DT, NN, JJ, NN, IN, NN, NNS, .]   
4  [NNP, NN, NN, NNP, NNP, VBD, NNP, ,, NNP, ,, J...   

                                                 Tag  \
0  [O, O, O, O, O, O, B-geo, O, O, O, O, O, B-geo...   
1  [B-gpe, O, O, O, O, O, O, O, O, O, O, O, O, O,...   
2  [O, O, B-tim, O, O, O, O, O, B-geo, O, O, O, O...   
3                  [O, O, O, O, O, O, O, O, O, O, O]   
4  [B-geo, O, O, B-per, I-per, O, B-tim, O, B-geo...   

                                            Word_idx  \
0  [26203, 28887, 29245, 19764, 5249, 29784, 8083...   
1  [34406, 30379, 28036, 27533, 27823, 34978, 199...   
2  [11030, 938, 11454, 30511, 6439, 13362, 2898, ...   
3  [27366, 7224, 28552, 31154, 6654, 25631, 709, ...   
4  [23564, 20418, 8520, 14516, 27641, 34003, 3357...   

                                             Tag_idx  
0  [16, 16, 16, 16, 16, 16, 13, 16, 16, 16, 16, 1...  
1  [11, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 1...  
2  [16, 16, 14, 16, 16, 16, 16, 16, 13, 16, 16, 1...  
3       [16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16]  
4  [13, 16, 16, 1, 3, 16, 14, 16, 13, 16, 11, 16,...

## 4. Pad sequences and split the dataset into train, test
Padding: The LSTM layers accept sequences of same length only. Therefore we will want to transform our list of token_sequences ('Word_idx') which is lists of integers into a matrix of shape (token_sequences, max_len). We can use any length as max_len. In this project we will be using length of the longest sequence as max_len. The sequences that are shorter than max_len are padded with a specified value at the end.
Import required packages

In [9]:
from sklearn.model_selection import train_test_split
from keras.utils import to_categorical, pad_sequences

In [10]:
def get_pad_train_test_val(data_group, data):
    n_token = len(list(set(data['Word'].to_list())))
    n_tag = len(list(set(data['Tag'].to_list())))
    
    tokens = data_group['Word_idx'].tolist()
    maxlen = max([len(s) for s in tokens])
    pad_tokens = pad_sequences(tokens, maxlen=maxlen, dtype='int32', padding='post', value= n_token - 1)
    
    tags = data_group['Tag_idx'].tolist()
    pad_tags = pad_sequences(tags, maxlen=maxlen, dtype='int32', padding='post', value= tag2idx["O"])
    
    n_tags = len(tag2idx)
    pad_tags = [to_categorical(i, num_classes=n_tags) for i in pad_tags]
    
    
    train_tokens, test_tokens, train_tags, test_tags = train_test_split(pad_tokens, pad_tags, test_size=0.1, train_size=0.9, random_state=2020)

    print(
        'train_tokens length:', len(train_tokens),
        '\ntest_tokens length:', len(test_tokens),
        '\ntrain_tags:', len(train_tags),
        '\ntest_tags:', len(test_tags)
    )
    
    return train_tokens, test_tokens, train_tags, test_tags

train_tokens, test_tokens, train_tags, test_tags = get_pad_train_test_val(data_group, data)

train_tokens length: 43163 
test_tokens length: 4796 
train_tags: 43163 
test_tags: 4796


In [11]:
for token, tag in zip(train_tokens[0], train_tags[0]):
    print('%s\t%s' % (token, tag))

26387	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
28036	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
16990	[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
3956	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
26993	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
28698	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
34978	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
9314	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
8930	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
22480	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
26508	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
15587	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
27533	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
17679	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
34035	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
21549	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
28879	[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]


In [12]:
import numpy as np
import tensorflow
from tensorflow.keras import Sequential, Model, Input
from tensorflow.keras.layers import LSTM, Embedding, Dense, TimeDistributed, Dropout, Bidirectional
from tensorflow.keras.utils import plot_model

# import keras as k

In [13]:
from numpy.random import seed
seed(1)
tensorflow.random.set_seed(2)

In [14]:
input_dim = len(list(set(data['Word'].to_list())))+1
output_dim = 32
input_length = max([len(s) for s in data_group['Word_idx'].tolist()])
n_tags = len(tag2idx)
print('input_dim: ', input_dim, '\noutput_dim: ', output_dim, '\ninput_length: ', input_length, '\nn_tags: ', n_tags)

input_dim:  35179 
output_dim:  32 
input_length:  104 
n_tags:  17


In [15]:
def get_bilstm_lstm_model():
    model = Sequential()

    # Add Embedding layer
    model.add(Embedding(input_dim=input_dim, output_dim=output_dim, input_length=input_length))

    # Add bidirectional LSTM
    model.add(Bidirectional(LSTM(units=output_dim, return_sequences=True, dropout=0.2, recurrent_dropout=0.2), merge_mode = 'concat'))

    # Add LSTM
    model.add(LSTM(units=output_dim, return_sequences=True, dropout=0.5, recurrent_dropout=0.5))

    # Add timeDistributed Layer
    model.add(TimeDistributed(Dense(n_tags, activation="relu")))

    #Optimiser 
    # adam = k.optimizers.Adam(lr=0.0005, beta_1=0.9, beta_2=0.999)

    # Compile model
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
    model.summary()
    
    return model

In [21]:

batch_size = 500
epochs = 5
verbose = 1
def train_model(X, y, model):
    loss = list()
    for _ in range(1):
        # fit model for one epoch on this sequence
        hist = model.fit(X, y, batch_size=batch_size, verbose=verbose, epochs=epochs, validation_split=0.2)
        loss.append(hist.history['loss'][0])
    return loss
model_bilstm_lstm = get_bilstm_lstm_model()
plot_model(model_bilstm_lstm)

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 104, 32)           1125728   
                                                                 
 bidirectional_2 (Bidirectio  (None, 104, 64)          16640     
 nal)                                                            
                                                                 
 lstm_5 (LSTM)               (None, 104, 32)           12416     
                                                                 
 time_distributed_2 (TimeDis  (None, 104, 17)          561       
 tributed)                                                       
                                                                 
Total params: 1,155,345
Trainable params: 1,155,345
Non-trainable params: 0
_________________________________________________________________
You must install pydot (`pip install pydot`)

In [22]:
results = pd.DataFrame()
results['with_add_lstm'] = train_model(train_tokens, np.array(train_tags), model_bilstm_lstm)

Epoch 1/5
70/70 [==============================] - 172s 2s/step - loss: 0.5265 - accuracy: 0.9565 - val_loss: 0.3587 - val_accuracy: 0.9679
Epoch 2/5
70/70 [==============================] - 163s 2s/step - loss: 0.3938 - accuracy: 0.9444 - val_loss: 0.3459 - val_accuracy: 0.9283
Epoch 3/5
70/70 [==============================] - 163s 2s/step - loss: 0.3121 - accuracy: 0.9607 - val_loss: 0.2785 - val_accuracy: 0.9679
Epoch 4/5
70/70 [==============================] - 161s 2s/step - loss: 0.2822 - accuracy: 0.9677 - val_loss: 0.2631 - val_accuracy: 0.9679
Epoch 5/5
70/70 [==============================] - 163s 2s/step - loss: 0.2707 - accuracy: 0.9677 - val_loss: 0.2532 - val_accuracy: 0.9679


After running 5 epochs with 500 batch size, the final accuracy was 0.9679. 

## spaCy entities

In [ ]:
import spacy
from spacy import displacy
nlp = spacy.load('en_core_web_sm')

In [ ]:
text = nlp(
    'Jim bought 300 shares of Acme Corp. in 2006. And producing an annotated block of text that \
    highlights the names of entities: [Jim]Person bought 300 shares of \
    [Acme Corp.]Organization in [2006]Time. In this example, a person name consisting \
    of one token, a two-token company name and a temporal expression have been detected \
    and classified.State-of-the-art NER systems for English produce near-human performance. \
    For example, the best system entering MUC-7 scored 93.39% of F-measure while human \
    annotators scored 97.60% and 96.95%.[1][2]')

In [ ]:
for w in text.ents:
    print(w.text, w.label_)

In [ ]:
displacy.render(text, style = 'ent', jupyter=True)

In [ ]:
spacy.explain('CARDINAL')